jena-arq/src/main/java/org/apache/jena/riot/resultset/rw/ResultSetWriterXML.java

```diff
@@ -253,84 +254,138 @@ public void binding(String varName, RDFNode node) {
             out.print(varName);
             out.println("\">");
             out.incIndent(INDENT);
-            printBindingValue(node);
+            printBindingValue(rdfNode);
             out.decIndent(INDENT);
             out.print("</");
             out.print(dfBinding);
             out.println(">");
         }
 
-        void printBindingValue(RDFNode node) {
-            if ( node == null ) {
+        private void printBindingValue(RDFNode rdfNode) {
+            if ( rdfNode == null ) {
                 // Unbound
                 out.print("<");
                 out.print(dfUnbound);
                 out.println("/>");
                 return;
             }
 
-            if ( node instanceof Literal ) {
-                printLiteral((Literal)node);
+            Node node = rdfNode.asNode();
+            printBindingValue(node);
+        }
+        
+        private void printBindingValue(Node node) {
+            if ( node == null )
+                return;
+        
+            if ( node.isLiteral() ) {
+                printLiteral(node);
                 return;
             }
 
-            if ( node instanceof Resource ) {
-                printResource((Resource)node);
+            if ( node.isURI() ) {
+                printURI(node);
+                return;
+            }
+            
+            if ( node.isBlank() ) {
+                printBlankNode(node);
+                return;
+            }
+            if ( node.isNodeTriple() ) {
+                printTripleTerm(node);
                 return;
             }
+            
+            if ( node.isNodeGraph() )
+                throw new UnsupportedOperationException("Graph terms");
 
-            Log.warn(this, "Unknown RDFNode type in result set: " + node.getClass());
+            Log.warn(this, "Unknown RDFNode type in result set: " + node);
         }
 
-        void printLiteral(Literal literal) {
+        private void printURI(Node nodeURI) {
+            String uri = nodeURI.getURI();
             out.print("<");
-            out.print(dfLiteral);
+            out.print(dfURI);
+            out.print(">");
+            out.print(xml_escape(uri));
+            out.print("</");
+            out.print(dfURI);
+            out.println(">");
+        }            
+            
+        private void printBlankNode(Node node) {
+            String label = bNodeMap.get(null, node);
+            // Comes with leading "_:"
+            label = label.substring(2);
+            out.print("<");
+            out.print(dfBNode);
+            out.print(">");
```

Comment:
The new methods `openTag`/`closeTag` cannot be used in `printBlankNode` too, as in `printTripleTerm`? (asking because I noticed that method also increments indentation, so maybe not supposed to be always used?)

Reply:
The XML write is very, very, _very_ old! I can't even track its history in git properly. 

The new `openTag`/`closeTag` made writing the nested structures for `printTripleTerm`.

I tried to limit the scope of changes and not do a whole-class clear-up, though I was sorely tempted to because the code is very un-DRY.